add vhost-user-vsock device emulation

stefano-garzarella opened the pull request:
This PR is based on https://github.com/rust-vmm/vhost-device/pull/7.

@harshanavkis implemented a _vhost-user-vsock_ device for a [GSoC 2021 project](https://summerofcode.withgoogle.com/archive/2021/projects/6126117680840704) mentored by me and @fidencio.

His PR was in a great state, I just rebased and completed the following items to be able to merge this project that could be very useful:
- updated some dependencies (e.g. vm-memory, etc.)
- replaced the deprecated `App::from_yaml()`
- fixed some new warnings from `clippy`
- addressed @stsquad comments (don't panic when `CONNECT 1234` is wrong, use [virtio-vsock 0.1.0](https://crates.io/crates/virtio-vsock) crate from @lauralt )
- avoided sending credit updates when the credit hasn't changed
- added `SPDX-License-Identifier` in each vsock file

stefano-garzarella commented:
v2:
- fixed more clippy warnings (I run the nightly clippy to be sure I cover also the new ones)
- fixed commit title length issue

TODO:
- [ ] extend test coverage (this PR drops the coverage by 20%, would be nice to extend the tests. I'll try to do my best in the next days, but if someone wants to help, it will be great)

stefano-garzarella commented:
v3:
- fixed another commit length issue
- reverted a change that I added by mistake in v2

TODO:
- [ ]  extend test coverage (this PR drops the coverage by 20%, would be nice to extend the tests. I'll try to do my best in the next days, but if someone wants to help, it will be great)

stefano-garzarella commented:
v4:
- added more test to reach ~70% of coverage (it was ~55% before the new tests).
- removed some unwrap() and fixed little issues discovered while writing tests

I hope the coverage is not that bad, and we can merge this. In the future we can try to increase it, adding more moks.

stsquad commented on vsock/src/thread_backend.rs:
```diff
@@ -0,0 +1,307 @@
+use virtio_vsock::packet::VsockPacket;
+use vm_memory::bitmap::BitmapSlice;
+
+// TODO: convert UnixStream to Arc<Mutex<UnixStream>>
```
Why not convert now? Can we have multiple threads now?

stefano-garzarella replied:
Nope, we have a single thread and I think that was an old TODO since `VsockThreadBackend` is supposed to be per-thread (IIUC).
Anyway, if we will move to multi-thread, I think we need to handle also other maps, so I would remove that one if you and  @harshanavkis agree.

stsquad replied:
fine by me

stsquad commented on vsock/src/main.rs:
```diff
@@ -0,0 +1,85 @@
+        let mut vring_workers = vsock_daemon.get_epoll_handlers();
+
+        if vring_workers.len() != vsock_backend.read().unwrap().threads.len() {
+            println!("Number of vring workers must be identical to number of backend threads");
```
Should we do something here instead of just notify the user? Can we use error!() here for reporting?

stefano-garzarella replied:
I guess it was here just for debug purpose, so I would remove it or just add an assert!().
VhostUserDaemon takes this information from `VhostUserVsockBackend::queues_per_thread`.
In this case `VhostUserVsockBackend` setup a single thread and a single element in `queues_per_thread`, so this check should be always false.

stefano-garzarella replied:
Or maybe I can add a test in main.c to check these things.

stsquad commented on vsock/src/main.rs:
```diff
@@ -0,0 +1,85 @@
+        }
+
+        match vsock_daemon.wait() {
+            Ok(()) => {
```
Use info! and warn! macros for reporting so we can follow the same standard reporting as the other daemons.

stefano-garzarella replied:
ack

stsquad commented on vsock/src/main.rs:
```diff
@@ -0,0 +1,85 @@
+            .unwrap()
+            .exit_event
+            .write(1)
+            .expect("Shutting down worker thread");
```
I'd argue this is superflous verbiage

stefano-garzarella replied:
Agree, using `unwrap`

stsquad commented on vsock/src/vhu_vsock_thread.rs:
```diff
@@ -0,0 +1,688 @@
+            epoll::Events::EPOLLIN,
+        )?;
+
+        // self.register_listener(stream_fd, BACKEND_EVENT);
```
superfluous commented out code?

stsquad commented on vsock/src/vhu_vsock_thread.rs:
```diff
@@ -0,0 +1,688 @@
+                self.process_rx_queue(vring)?;
+                if !vring.enable_notification().unwrap() {
+                    break;
+                }
```
please remove commented out code

stefano-garzarella replied:
ack

stsquad reviewed (changes requested):
I've tested the daemon on my setup and everything looks pretty good to me. I have added a few comments about removing commented out code and cleaning up the println -> info!,warn! and debug! macros. I think the remaining TODOs and further increasing the coverage can be done once this is merged.

stefano-garzarella commented:
v5:
- replaced all `println!()` with `dbg!()`, `info!()`, or `error!()`
- initialized `env_logger` in main()
- removed old comments, useless code, and code commented
- added test in main.rs to check that `VhostUserDaemon`'s workers and `VhostUserVsockBackend`'s threads are the same (they should be always the same if the code is correct)

stefano-garzarella commented:
@stsquad thanks for testing and reviews! I just addressed all your comments and pushed the new version

stefano-garzarella commented:
> I will start looking into this pull request in the coming days. Given its sheer size I expect the review process to take several days. I will clearly let you know when I am done.

@mathieupoirier thank you very much! Yep, take your time. I have taken up Harsha's work, so some part may be unknown to me too ;-)

vireshk commented on vsock/src/vhu_vsock.rs:
```diff
@@ -0,0 +1,488 @@
+}
+
+pub struct VhostUserVsockBackend {
+    guest_cid: __u64,
```
Why __u64 here ?

Since we need to set this only once and then just send it in Little endian format, we can make it vm_memory::Le64 instead and set it as:

guest_cid = From::from(u64-value);

This is what's done in GPIO crate.

stefano-garzarella replied:
yep, make sense to me!

vireshk commented on vsock/src/thread_backend.rs:
```diff
@@ -0,0 +1,306 @@
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+#![deny(missing_docs)]
```
still required ?

stefano-garzarella replied:
I don't know, I'll check!

vireshk commented on vsock/src/thread_backend.rs:
```diff
@@ -0,0 +1,306 @@
+use virtio_vsock::packet::VsockPacket;
+use vm_memory::bitmap::BitmapSlice;
+
+pub struct VsockThreadBackend {
```
pub(crate) should be enough ? I think that's true for most of the stuff here as this is a binary crate and most of the stuff shouldn't be visible out of this crate. I do see some violation of this in existing crates, will add an issue for that.

stefano-garzarella replied:
okay, I'll change

vireshk commented on vsock/src/thread_backend.rs:
```diff
@@ -0,0 +1,306 @@
+    }
+
+    /// Checks if there are pending rx requests in the backend
+    /// rxq.
```
Single line should be enough for this comment.

stefano-garzarella replied:
Agree

vireshk commented on vsock/src/thread_backend.rs:
```diff
@@ -0,0 +1,306 @@
+            Some(cmk) => cmk,
+            None => {
+                return Err(Error::EmptyBackendRxQ);
+            }
```
What about:

        let key = match self.backend_rxq.pop_front().map_err(|_| Error::EmptyBackendRxQ)?;

stefano-garzarella replied:
Yep, I'll change

stefano-garzarella replied:
I just realized that the value returned by `pop_front()` is `std::option::Option` and `.map_err` is not available.
Is there any other nice way? otherwise I'll leave it like this.

vireshk replied:
Maybe somthing like this ?

let key = self.backend_rxq.pop_front().ok_or(Error::EmptyBackendRxQ)?;

vireshk commented on vsock/src/thread_backend.rs:
```diff
@@ -0,0 +1,306 @@
+                .set_buf_alloc(0)
+                .set_fwd_cnt(0);
+
+            return Ok(());
```
drop this and use if/else instead ?

vireshk replied:
its okay then, no issues.

vireshk reviewed (changes requested):
About code coverage, I looked at kcov_output/ to see what all is missed and there are things we can easily cover I think. For example main.rs::start_backend_server() and a lot of other code in vhu_vsock_thread.rs which is not covered as of now in tests.

vireshk commented on vsock/src/vhu_vsock.rs:
```diff
@@ -0,0 +1,488 @@
+    fn update_memory(
+        &mut self,
+        atomic_mem: GuestMemoryAtomic<GuestMemoryMmap>,
+    ) -> result::Result<(), io::Error> {
```
Instead use std::io::Result directly.

    ) -> std::io::Result<()> {

For I2c/gpio, I have aliased that as IoResult (in a new pull request).

same applies to handle_event and elsewhere.

stefano-garzarella replied:
Can we do this as a follow-up PR? (maybe opening an issue and when someone has time, we can fix)

stefano-garzarella replied:
Fixed, it was just in the tests.

stefano-garzarella commented on vsock/src/main.rs:
```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
```
I'm not the original author of this code, so I just added the SPDX-License-Identifier following what Harsha specified in his Cargo.toml.

I'm open to change on Apache-2.0 only, but we need an approval from @harshanavkis.

harshanavkis replied:
Fine by me.

stefano-garzarella replied:
@harshanavkis thanks for your ACK, I'll prepare a PR to re-license vhost-user-vsock as Apache-2.0 only.

stefano-garzarella commented on vsock/src/main.rs:
```diff
@@ -0,0 +1,111 @@
+fn main() {
+    env_logger::init();
+
+    let vsock_config = VsockConfig::try_from(VsockArgs::parse()).unwrap();
```
This was by Harsha, I'm open to move VsockConfig and VsockArgs in main, if you prefer.

vireshk replied:
That would be nice. Thanks.

stefano-garzarella replied:
`VsockConfig` is also used by `VhostUserVsockBackend` defined in vhu_vsock.rs
I think the original idea was to export part of the crate also as a library (in the future it will be nice, since I have a couple of use cases).
So for now I'll move only VsockArgs in main.rs, since it is really only needed by the binary.

stefano-garzarella commented on vsock/src/main.rs:
```diff
@@ -0,0 +1,111 @@
+/// vhost-user-vsock backend server.
+pub(crate) fn start_backend_server(vsock_config: VsockConfig) {
+    loop {
+        let vsock_backend = Arc::new(RwLock::new(
```
I'll try to fix.

stefano-garzarella commented on vsock/src/main.rs:
```diff
@@ -0,0 +1,111 @@
+        let listener = Listener::new(vsock_config.get_socket_path(), true).unwrap();
+
+        let mut vsock_daemon = VhostUserDaemon::new(
+            String::from("vhost-user-vsock"),
```
I think Harsha followed what we did in QEMU where we have `vhost-user-blk`, `vhost-user-gpu`, `vhost-user-scsi`, `vhost-user-input`, etc. Honestly I would have done the same, at the most we could use `vhost-user-X-device`, though, I don't think _daemon_ adds much.

vireshk replied:
Maybe then we should convert existing backends to use this nomenclature. I just want them to be consistent and yes daemon name doesn't matter much otherwise.

vireshk replied:
Maybe just keep vhost-user-vsock for now ? We can discuss later in a separate pull request on how to name all of these.

stefano-garzarella replied:
Ack

stefano-garzarella commented on vsock/src/main.rs:
```diff
@@ -0,0 +1,111 @@
+use vhost::{vhost_user, vhost_user::Listener};
+use vhost_user_backend::VhostUserDaemon;
+use vhu_vsock::{VhostUserVsockBackend, VsockArgs, VsockConfig};
+use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
```
> About code coverage, I looked at kcov_output/ to see what all is missed and there are things we can easily cover I think. For example main.rs::start_backend_server() and a lot of other code in vhu_vsock_thread.rs which is not covered as of now in tests.

I agree, but can we postpone?
I already spent a lot of time to rebase Harsha's work, so my plan will be to fix the simple things and open issues to postpone the rest. (maybe some student can start with tests or refactoring, or I can do it when I'll have more time)
Is that okay?

stefano-garzarella replied:
Ack

stefano-garzarella replied:
In i2c and gpio I see `use log...` in the first group. IIUC `log` is an external crate. Should it be in the second group?

vireshk commented:
> I agree, but can we postpone? I already spent a lot of time to rebase Harsha's work, so my plan will be to fix the simple things and open issues to postpone the rest. (maybe some student can start with tests or refactoring, or I can do it when I'll have more time) Is that okay?

It is always better / easier to fix things at review time as they are lost otherwise. But I do get what you are saying here.

Please do whatever changes you can to the current patches, based on current comments, and we will get it merged. I can also try to spend some cycles on this then and raise issues too.

stefano-garzarella commented:
> It is always better / easier to fix things at review time as they are lost otherwise. But I do get what you are saying here.
> 

I totally agree with you, but since I'm not the author, it's like doing a review every time to see if what I'm testing or changing makes sense :-)

> Please do whatever changes you can to the current patches, based on current comments, and we will get it merged. I can also try to spend some cycles on this then and raise issues too.

Thank you for understanding, I will do my best.
If you need help maintaining this workspace, I can allocate some time to help and add myself to `CODEOWNERS`.

stefano-garzarella commented on vsock/src/vhu_vsock.rs:
```diff
@@ -0,0 +1,488 @@
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+//! vhost-user-vsock device crate
```
I don't know, I'll remove

stefano-garzarella commented on vsock/src/vhu_vsock.rs:
```diff
@@ -0,0 +1,488 @@
+            guest_cid: vsock_config.get_guest_cid(),
+            threads,
+            queues_per_thread,
+            exit_event: EventFd::new(EFD_NONBLOCK).expect("Creating exit eventfd"),
```
yep

stefano-garzarella commented on vsock/src/vhu_vsock.rs:
```diff
@@ -0,0 +1,488 @@
+        let vring_tx = &vrings[1];
+
+        if evset == EventSet::OUT {
+            dbg!("received epollout");
```
I think so, I'll remove

stefano-garzarella commented:
v6:
 - addressed most of @vireshk requests
   - reordered `use` declarations
   - removed "vsock_" prefix where is redundant
   - moved VsockArgs in main.rs
   - used pub(crate) for types used only by our binary
   - and other small changes merged on previous commits

TODO:
- [ ] extend test coverage (I prefer as a follow up in a new PR, I'll open an issue to track)

stefano-garzarella commented:
@vireshk @stsquad @mathieupoirier I just pressed the button to request Alex's review and GH messed up other requests. Sorry about that ;-)

mathieupoirier commented:
I see that Viresh has beaten me to the punch on reviewing this pull request.  There is no need to do things twice, especially with a patchset this big, and as such will not continue with this work.

vireshk commented:
> I see that Viresh has beaten me to the punch on reviewing this pull request. There is no need to do things twice, especially with a patchset this big, and as such will not continue with this work.

You can approve it for now, I have already done so and we will continue to modify it to make it similar to other crates and fix any more issues we find later.
